File: Python/OneLang/One/Transforms/InferTypesPlugins/InferReturnType.py
from OneLangStdLib import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Statements as stats
import OneLang.One.ErrorManager as errorMan
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.Interfaces as ints

class ReturnTypeInferer:
    def __init__(self, error_man):
        self.returns_null = False
        self.throws = False
        self.return_types = []
        self.error_man = error_man
    
    def add_return(self, return_value):
        if isinstance(return_value, exprs.NullLiteral):
            self.returns_null = True
            return
        
        return_type = return_value.actual_type
        if return_type == None:
            raise Error("Return type cannot be null")
        
        if not ArrayHelper.some(lambda x: astTypes.TypeHelper.equals(x, return_type), self.return_types):
            self.return_types.append(return_type)
    
    def finish(self, declared_type, error_context, async_type):
        inferred_type = None
        
        if len(self.return_types) == 0:
            if self.throws:
                inferred_type = declared_type or astTypes.VoidType.instance
            elif self.returns_null:
                if declared_type != None:
                    inferred_type = declared_type
                else:
                    self.error_man.throw(f'''{error_context} returns only null and it has no declared return type!''')
            else:
                inferred_type = astTypes.VoidType.instance
        elif len(self.return_types) == 1:
            inferred_type = self.return_types[0]
        elif declared_type != None and ArrayHelper.every(lambda x, i: astTypes.TypeHelper.is_assignable_to(x, declared_type), self.return_types):
            inferred_type = declared_type
        else:
            self.error_man.throw(f'''{error_context} returns different types: {", ".join(list(map(lambda x: x.repr(), self.return_types)))}''')
            inferred_type = astTypes.AnyType.instance
        
        check_type = declared_type
        if check_type != None and async_type != None and isinstance(check_type, astTypes.ClassType) and check_type.decl == async_type.decl:
            check_type = check_type.type_arguments[0]
        
        if check_type != None and not astTypes.TypeHelper.is_assignable_to(inferred_type, check_type):
            self.error_man.throw(f'''{error_context} returns different type ({inferred_type.repr()}) than expected {check_type.repr()}''')
        
        self.return_types = None
        return declared_type if declared_type != None else inferred_type

class InferReturnType(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        self.return_type_infer = []
        super().__init__("InferReturnType")
    
    def get_current(self):
        return self.return_type_infer[len(self.return_type_infer) - 1]
    
    def start(self):
        self.return_type_infer.append(ReturnTypeInferer(self.error_man))
    
    def finish(self, declared_type, error_context, async_type):
        return self.return_type_infer.pop().finish(declared_type, error_context, async_type)
    
    def handle_statement(self, stmt):
        if len(self.return_type_infer) == 0:
            return False
        if isinstance(stmt, stats.ReturnStatement) and stmt.expression != None:
            self.main.process_statement(stmt)
            self.get_current().add_return(stmt.expression)
            return True
        elif isinstance(stmt, stats.ThrowStatement):
            self.get_current().throws = True
            return False
        else:
            return False
    
    def handle_lambda(self, lambda_):
        self.start()
        self.main.process_lambda(lambda_)
        lambda_.returns = self.finish(lambda_.returns, "Lambda", None)
        lambda_.set_actual_type(astTypes.LambdaType(lambda_.parameters, lambda_.returns), False, True)
        return True
    
    def handle_method(self, method):
        if isinstance(method, types.Method) and method.body != None:
            self.start()
            self.main.process_method_base(method)
            method.returns = self.finish(method.returns, f'''Method "{method.name}"''', self.main.current_file.literal_types.promise if method.async_ else None)
            return True
        else:
            return False
    
    def handle_property(self, prop):
        self.main.process_variable(prop)
        
        if prop.getter != None:
            self.start()
            self.main.process_block(prop.getter)
            prop.type = self.finish(prop.type, f'''Property "{prop.name}" getter''', None)
        
        if prop.setter != None:
            self.start()
            self.main.process_block(prop.setter)
            self.finish(astTypes.VoidType.instance, f'''Property "{prop.name}" setter''', None)
        
        return True

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/InferReturnType.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\InferReturnType;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Types\Property;
use OneLang\One\Ast\Types\Lambda;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Types\IMethodBase;
use OneLang\One\Ast\AstTypes\VoidType;
use OneLang\One\Ast\AstTypes\AnyType;
use OneLang\One\Ast\AstTypes\LambdaType;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\TypeHelper;
use OneLang\One\Ast\Statements\Statement;
use OneLang\One\Ast\Statements\ReturnStatement;
use OneLang\One\Ast\Statements\ThrowStatement;
use OneLang\One\ErrorManager\ErrorManager;
use OneLang\One\Ast\Expressions\NullLiteral;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Interfaces\IType;

class ReturnTypeInferer {
    public $returnsNull = false;
    public $throws = false;
    public $returnTypes;
    public $errorMan;
    
    function __construct($errorMan) {
        $this->errorMan = $errorMan;
        $this->returnTypes = array();
    }
    
    function addReturn($returnValue) {
        if ($returnValue instanceof NullLiteral) {
            $this->returnsNull = true;
            return;
        }
        
        $returnType = $returnValue->actualType;
        if ($returnType === null)
            throw new \OneLang\Core\Error("Return type cannot be null");
        
        if (!\OneLang\Core\ArrayHelper::some($this->returnTypes, function ($x) use ($returnType) { return TypeHelper::equals($x, $returnType); }))
            $this->returnTypes[] = $returnType;
    }
    
    function finish($declaredType, $errorContext, $asyncType) {
        $inferredType = null;
        
        if (count($this->returnTypes) === 0) {
            if ($this->throws)
                $inferredType = $declaredType ?? VoidType::$instance;
            else if ($this->returnsNull) {
                if ($declaredType !== null)
                    $inferredType = $declaredType;
                else
                    $this->errorMan->throw($errorContext . " returns only null and it has no declared return type!");
            }
            else
                $inferredType = VoidType::$instance;
        }
        else if (count($this->returnTypes) === 1)
            $inferredType = $this->returnTypes[0];
        else if ($declaredType !== null && \OneLang\Core\ArrayHelper::every($this->returnTypes, function ($x, $i) use ($declaredType) { return TypeHelper::isAssignableTo($x, $declaredType); }))
            $inferredType = $declaredType;
        else {
            $this->errorMan->throw($errorContext . " returns different types: " . implode(", ", array_map(function ($x) { return $x->repr(); }, $this->returnTypes)));
            $inferredType = AnyType::$instance;
        }
        
        $checkType = $declaredType;
        if ($checkType !== null && $asyncType !== null && $checkType instanceof ClassType && $checkType->decl === $asyncType->decl)
            $checkType = $checkType->typeArguments[0];
        
        if ($checkType !== null && !TypeHelper::isAssignableTo($inferredType, $checkType))
            $this->errorMan->throw($errorContext . " returns different type (" . $inferredType->repr() . ") than expected " . $checkType->repr());
        
        $this->returnTypes = null;
        return $declaredType !== null ? $declaredType : $inferredType;
    }
}

class InferReturnType extends InferTypesPlugin {
    public $returnTypeInfer;
    
    function get_current() {
        return $this->returnTypeInfer[count($this->returnTypeInfer) - 1];
    }
    
    function __construct() {
        parent::__construct("InferReturnType");
        $this->returnTypeInfer = array();
    }
    
    function start() {
        $this->returnTypeInfer[] = new ReturnTypeInferer($this->errorMan);
    }
    
    function finish($declaredType, $errorContext, $asyncType) {
        return array_pop($this->returnTypeInfer)->finish($declaredType, $errorContext, $asyncType);
    }
    
    function handleStatement($stmt) {
        if (count($this->returnTypeInfer) === 0)
            return false;
        if ($stmt instanceof ReturnStatement && $stmt->expression !== null) {
            $this->main->processStatement($stmt);
            $this->get_current()->addReturn($stmt->expression);
            return true;
        }
        else if ($stmt instanceof ThrowStatement) {
            $this->get_current()->throws = true;
            return false;
        }
        else
            return false;
    }
    
    function handleLambda($lambda) {
        $this->start();
        $this->main->processLambda($lambda);
        $lambda->returns = $this->finish($lambda->returns, "Lambda", null);
        $lambda->setActualType(new LambdaType($lambda->parameters, $lambda->returns), false, true);
        return true;
    }
    
    function handleMethod($method) {
        if ($method instanceof Method && $method->body !== null) {
            $this->start();
            $this->main->processMethodBase($method);
            $method->returns = $this->finish($method->returns, "Method \"" . $method->name . "\"", $method->async ? $this->main->currentFile->literalTypes->promise : null);
            return true;
        }
        else
            return false;
    }
    
    function handleProperty($prop) {
        $this->main->processVariable($prop);
        
        if ($prop->getter !== null) {
            $this->start();
            $this->main->processBlock($prop->getter);
            $prop->type = $this->finish($prop->type, "Property \"" . $prop->name . "\" getter", null);
        }
        
        if ($prop->setter !== null) {
            $this->start();
            $this->main->processBlock($prop->setter);
            $this->finish(VoidType::$instance, "Property \"" . $prop->name . "\" setter", null);
        }
        
        return true;
    }
}
